src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/share/service/IntersectionComputer.java

```diff
@@ -131,6 +132,61 @@ public static boolean isProtectedPathOverlaps(
         return !Sets.intersection(primaryEdges, protectedEdges).isEmpty();
     }
 
+    /**
+     * Calculates intersection (shared part) among the paths starting from the source endpoint.
+     *
+     * @param paths the paths to examine.
+     * @return the overlapping path segments.
+     */
+    public static List<PathSegment> calculatePathIntersectionFromSource(Collection<FlowPath> paths) {
+        if (paths.size() < 2) {
+            throw new IllegalArgumentException("At least 2 paths must be provided");
+        }
+
+        List<PathSegment>[] pathSegments = paths.stream().map(FlowPath::getSegments).toArray(List[]::new);
+
+        int minLength = pathSegments[0].size();
+        for (List<?> segments : pathSegments) {
+            if (segments.isEmpty()) {
+                throw new IllegalArgumentException("All paths mustn't be empty");
+            }
+            if (minLength > segments.size()) {
+                minLength = segments.size();
+            }
+        }
+        SwitchId source = pathSegments[0].get(0).getSrcSwitchId();
+        for (List<PathSegment> path : pathSegments) {
+            if (!path.get(0).getSrcSwitchId().equals(source)) {
+                throw new IllegalArgumentException("All paths must have the same source endpoint");
+            }
+        }
+        List<PathSegment> result = new ArrayList<>();
+        for (int i = 0; i < minLength; i++) {
```

Comment:
`minLength` is actually `maxLength`.

As for me make iterator for each provided path read one entry from each iterator and and continue while all entries are equal will be simpler.

Reply:
Refactored.